[62341] Correctly convert dates when timezone is negative in datepicker

cbliard opened the pull request:
# Ticket

https://community.openproject.org/wp/62341



<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What are you trying to accomplish?

The dates are off by 1 or 2 days when in a negative timezone, both in the calendar view and the date fields.

## Screenshots

Video in to linked ticket.

# What approach did you choose and why?

flatpickr is using local dates at midnight internally and in its callback functions. In datepicker component and stimulus controller we are using UTC dates at midnight.

And the timezone service is assuming local time when converting to iso dates.

Usingtimezones with positive offset is ok, but with negative offset, the conversions and comparisons do not work properly.

Fixed it by explicitly converting from utc date to iso string, and using iso date strings when setting flatpickr dates, avoiding any issues.

It could be a good idea to use iso date strings everywhere as much as possible instead of always jumping back and forth from dates to iso string.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
[62341] Correctly convert dates when timezone is negative in datepicker

flatpickr is using local dates at midnight internally and in its
callback functions. In datepicker component and stimulus controller we
are using UTC dates at midnight.

And the timezone service is assuming local time when converting to iso dates.

Fixed it by explcitly converting from utc date to iso string, and using
iso date strings when setting flatpickr dates, avoiding any issues.

It could be a good idea to use iso date strings everywhere as much as
possible instead of always jumping back and forth from dates to iso
string.

klaustopher commented on spec/support/cuprite_setup.rb:
```diff
@@ -108,7 +108,10 @@ def register_better_cuprite(language, name: :"better_cuprite_#{language}")
   config.around(:each, :js, type: :feature) do |example|
     # Skip if driver is explicitly requested
-    next if example.metadata[:driver]
+    if example.metadata[:driver]
+      example.run
+      next
+    end
```
So, since before we did not call `example.run` in the around filter we did never run the test and now we are running it, but should we not set the `Capybara.javascript_driver` as well? Or is that automatically done by the `driver:` metadata?

klaustopher replied:
<img width="862" alt="Bildschirmfoto 2025-03-20 um 11 44 58" src="https://github.com/user-attachments/assets/086e6b9e-012c-4bd1-8f7b-157131f63844" />

Setting the driver is part of capybara itself, so yes ... This does what it says on the tin. Good catch!

cbliard replied:
It's automatically done by the `driver:` metadata. It's handled by Capybara.

klaustopher reviewed (approved):
approved in general, would be nice to clarify the one comment I added

cbliard committed:
Update test with single date mode

This date was previously skipped because of the way the `:driver` metadata was used.

It was failing because it was not updated to the new single mode, and
the due date was not active, so clicking the date in the calendar was
changing the start date, whereas the test expected the due date to be
set.

The test has been fixed by enabling the due date first. Some additional
checks have been added as well.
